### PR TITLE
⬆️ Upgrade to Fotan by applying CosmJS

### PR DESCRIPTION
**Pull request by WeiJun0827**

### CODE REVIEW

File: util/cosmos.js [hunk 1]
```diff
@@ -1,43 +1,29 @@
...
   let notFoundOnce = false;
+  const client = await StargateClient.connect(COSMOS_ENDPOINT);
   while (!tx) {
     /* eslint-disable no-await-in-loop */
     await timeout(1000);
     try {
-      const { data } = await api.get(`/${txHash}`);
-      if (data && data.height) {
+      const data = await client.getTx(txHash);
+      if (data) {
```
Comment (WeiJun0827) [inline util/cosmos.js:15]: Note: `data` will either be `null` or type `IndexedTx`, ref [here](https://github.com/cosmos/cosmjs/blob/main/packages/stargate/src/stargateclient.ts#L286).

File: util/cosmos.js [hunk 2]
```diff
@@ -1,43 +1,29 @@
...
         ({ tx } = data);
-        const {
-          code,
-          logs: [{ success = false } = {}] = [],
-        } = data;
-        const isFailed = (code && code !== '0') || !success;
-        if (isFailed) throw new Error(code);
+        const { code } = data;
+        if (code) throw new Error(code);
```
Comment (WeiJun0827) [inline util/cosmos.js:18]: In stargate, `logs` is no longer available in `data`. So I check if `code` is not 0 only.
(See type `IndexedTx` diff in [launchpad](https://github.com/cosmos/cosmjs/blob/main/packages/launchpad/src/cosmosclient.ts#L104) & [stargate](https://github.com/cosmos/cosmjs/blob/main/packages/stargate/src/stargateclient.ts#L62))

File: server/util/cosmos.js [hunk 3]
```diff
@@ -26,34 +26,41 @@ import {
...
-    tendermint34Client,
-    setupAuthExtension,
-    setupBankExtension,
-  );
-  signingClient = await SigningStargateClient.connectWithSigner(cosmosRPCEndpoint, wallet);
+const getCosmos = (() => {
+  let cosmos;
+  return async () => {
+    if (!cosmos) {
+      const wallet = await DirectSecp256k1Wallet.fromKey(COSMOS_PRIVATE_KEY);
```
Comment (WeiJun0827) [inline server/util/cosmos.js:33]: Concept source of this sophisticated async lazy initializer [here](https://stackoverflow.com/questions/58796320/what-is-a-canonical-safe-way-of-async-lazy-initialization-in-javascript/58796336#58796336).

File: server/util/cosmos.js [hunk 4]
```diff
@@ -13,225 +19,122 @@ import { PUBSUB_TOPIC_MISC } from '../constant';
...
-    try {
-      txHash = await sendTransaction(signedTx);
-    } catch (err) {
+  try {
+    txHash = await sendTransaction(signedTx);
+  } catch (err) {
+    // eslint-disable-next-line no-console
+    console.error(err);
+    const { message } = err;
+    if (message && message.includes('code 32')) {
```
Comment (williamchong) [inline server/util/cosmos.js:122]: The error returned should be a JSON object that is parsable with `code == 32`

File: server/util/cosmos.js [hunk 5]
```diff
@@ -13,225 +19,122 @@ import { PUBSUB_TOPIC_MISC } from '../constant';
...
-      const { sequence: seq } = await getAccountInfo(signer.cosmosAddress);
-      pendingCount = Number(seq);
-      signedTx = signTransaction({
-        signer,
-        accNum,
-        stdTx,
-        sequence: pendingCount,
-      });
+    if (!txHash) {
+      const { sequence: seq2 } = await getAccountInfo(delegatorAddress);
```
Comment (williamchong) [inline server/util/cosmos.js:133]: Can we only refetch sequence when error is code:32?

File: server/util/cosmos.js [hunk 6]
```diff
@@ -13,225 +19,122 @@ import { PUBSUB_TOPIC_MISC } from '../constant';
...
-    if (res.data.code) {
-      throw new Error(res.data.raw_log);
-    }
-    return res.data.txhash;
-  } catch (err) {
-    // eslint-disable-next-line no-console
-    console.error(err);
-    if (err && err.response && err.response.data) {
-      if (err.response.data.error
-        && err.response.data.error.includes('Tx already exists')) {
```
Comment (williamchong) [inline server/util/cosmos.js:127]: Can we handle the tx exists case in the new code?
Like https://github.com/likecoin/likecoin-jedi/pull/51/files#diff-889cd002a3740ec9ed5df105b2d50359050ec2772898562502a3ddf4d3ea1180R289

**Review by williamchong — CHANGES_REQUESTED** [review]

File: server/util/cosmos.js (hunk 6 above)
Comment (WeiJun0827) [reply server/util/cosmos.js:127]: Curious about in what circumstance will we have a existed tx?

Comment (WeiJun0827) [reply server/util/cosmos.js:127]: And will we prefer to use the same way to manually calculate the `txHash` and return?

Comment (williamchong) [reply server/util/cosmos.js:127]: When the transaction was submitted, but somehow the txHash was not stored into database.
In this case the transaction will be retried, but the same transaction already exists in the node's txpool, an tx exists error would be thrown

Comment (williamchong) [reply server/util/cosmos.js:127]: Only when the returned error is `transaction already exists` we would want to calculate the hash offline

Comment (WeiJun0827) [reply server/util/cosmos.js:127]: Updated

**Review by williamchong — APPROVED** [review]

**Pull request merged**